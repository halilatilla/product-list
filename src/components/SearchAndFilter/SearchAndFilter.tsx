import { FC, useEffect } from 'react'

import { useAppSelector, useAppDispatch, setFilteredProducts } from '@src/store'
import { useFuseSearch } from '@src/hooks'
import { IProduct } from '@src/types'
import { filterByKeys } from '@src/constants'
import { Search } from '@src/components'

interface Props {
  className?: string
  products: IProduct[]
}

const SearchAndFilter: FC<Props> = ({ products }) => {
  const { filterBy, searchTerm } = useAppSelector((state) => state.filter)
  const dispatch = useAppDispatch()
  const { fuse } = useFuseSearch([filterByKeys.COLOR, filterByKeys.BRAND, filterByKeys.TITLE])

  useEffect(() => {
    if (filterBy.brand === '' && filterBy.color === '') {
      dispatch(setFilteredProducts(products))
      return
    }

    if (filterBy.brand !== '' && filterBy.color !== '') {
      dispatch(
        setFilteredProducts(
          fuse.search({ $and: [{ brand: filterBy.brand }, { color: filterBy.color }] }).map((result) => result.item),
        ),
      )
      return
    }

    dispatch(
      setFilteredProducts(
        fuse.search({ $or: [{ brand: filterBy.brand }, { color: filterBy.color }] }).map((result) => result.item),
      ),
    )
  }, [filterBy])

  useEffect(() => {
    if (searchTerm.length > 2) {
      dispatch(setFilteredProducts(fuse.search(searchTerm).map((result) => result.item)))
      return
    }

    if (filterBy.brand === '' && filterBy.color === '' && searchTerm.length < 2) {
      dispatch(setFilteredProducts(products))
    }
  }, [searchTerm])

  return <Search placeholder="25 milyon’dan fazla ürün içerisinde ara" />
}

export default SearchAndFilter
