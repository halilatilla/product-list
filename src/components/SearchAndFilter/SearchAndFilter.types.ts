import { IProduct } from '@src/types'

export default interface ISearchAndFilter {
  className?: string
  products: IProduct[]
}
