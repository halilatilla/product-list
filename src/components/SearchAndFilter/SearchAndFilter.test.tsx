import { render } from '@src/testUtils'

import SearchAndFilter from './SearchAndFilter'

describe('SearchAndFilter', () => {
  const PROPS = {
    products: [],
  }

  it.only('renders correctly', () => {
    const { container } = render(<SearchAndFilter {...PROPS} />)
    expect(container).toMatchSnapshot()
  })
})
