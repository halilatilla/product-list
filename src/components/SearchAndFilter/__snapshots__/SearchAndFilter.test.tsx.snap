// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchAndFilter renders correctly 1`] = `
<div>
  <input
    class="search"
    placeholder="25 milyon’dan fazla ürün içerisinde ara"
    type="search"
  />
</div>
`;
